libflux/src/semantic/parser/mod.rs

```diff
@@ -0,0 +1,1755 @@
+        Some(letter) if letter == '}' => Some(TokenType::RIGHTCURLYBRAC), 
+        Some(letter) if letter == '[' => Some(TokenType::LEFTSQUAREBRAC), 
+        Some(letter) if letter == ']' => Some(TokenType::RIGHTSQUAREBRAC), 
+        Some(letter) if letter == '(' => Some(TokenType::LEFTPAREN), 
+        Some(letter) if letter == ')' => Some(TokenType::RIGHTPAREN), 
+        Some(letter) if letter == '?' => Some(TokenType::QUESTIONMARK), 
+        Some(letter) if letter == ':' => Some(TokenType::COLON), 
+        Some(letter) if letter == ',' => Some(TokenType::COMMA), 
+        Some(letter) if letter == '+' => Some(TokenType::PLUS), 
+        Some(letter) if letter == '<' => {
+            match lexer.next() {
+                Some('-') => Some(TokenType::PIPE),
+                _ => Some(TokenType::ERROR), 
+            }
+        },
+        Some(letter) if letter == '-' => {
+            match lexer.next() {
+                Some('>') => Some(TokenType::ARROW), 
+                _ => Some(TokenType::ERROR), 
+            }
+        },
+        Some(_) => Some(TokenType::ERROR), 
+        _ => Some(TokenType::EOF), 
+    }
+}
+
+/** PARSER */
+
+struct Parser<'a> {
+    tokens: Peekable<Iter<'a, Token>>,
+}
+
+// parse_tokens is a testing function for directly parsing a vector of tokens. 
+// If parsing a &str polytype, use the Parse function instead
+fn parse_tokens(tokens: Vec<Token>) -> Result<PolyType, &'static str> {
+    let mut parser = Parser{ tokens: tokens.iter().peekable() };
+    parser.parse_polytype()
+}
+
+// Parse passes the source text through the Lexer, receives tokens and removes
+// whitespace. Its then initializes parsing and returns a PolyType AST.
+// This is the only function meant to be accessed by the end user. It handles 
+// both lexing and parsing of string polytypes. 
+pub fn Parse(source: &str) -> Result<PolyType, &'static str> {
+    let tokens = Lex(source); 
+    let mut parser = Parser{ tokens: tokens.iter().peekable() }; 
+    parser.parse_polytype()
+}
+
+impl Parser<'_> {
+    // next grabs the next token using the Iter()'s next method and unpacks
+    // the value if there are still tokens to parse
+    fn next(& mut self) -> Token {
+        match self.tokens.next() {
+            Some(token) => (*token).clone(),
+            None => Token {
+                token_type: TokenType::EOF,
+                text: None
+            },
+        }
+    }
+    
+    // peek returns a preview of the next Token using Iter()'s peek method
+    // and unpacks the value if there are still tokens to parse
+    fn peek(& mut self) -> Token {
+        match self.tokens.peek() {
+            Some(token) => (**token).clone(),
+            None => Token {
+                token_type: TokenType::EOF,
+                text: None
+            },
+        }
+    }
+
+
+    /*Production rules for each of the following methods can be found in the accompanying grammar.md file. 
```

Comment:
nit, can you remove this comment, or make it a doc comment over `parse_polytype` just to keep things consistent?